validator_client/src/config.rs

```diff
@@ -1,32 +1,131 @@
+use bincode;
+use bls::Keypair;
+use clap::ArgMatches;
 use std::fs;
+use std::fs::File;
+use std::io::{Error, ErrorKind};
 use std::path::PathBuf;
 use types::ChainSpec;
 
 /// Stores the core configuration for this validator instance.
 #[derive(Clone)]
-pub struct ClientConfig {
+pub struct ValidatorClientConfig {
+    /// The data directory, which stores all validator databases
     pub data_dir: PathBuf,
+    /// The directory where the individual validator configuration directories are stored.
+    pub validator_dir: PathBuf,
+    /// The server at which the Beacon Node can be contacted
     pub server: String,
+    /// The chain specification that we are connecting to
     pub spec: ChainSpec,
 }
 
-const DEFAULT_LIGHTHOUSE_DIR: &str = ".lighthouse-validators";
+const DEFAULT_VALIDATOR_DATADIR: &str = ".lighthouse-validator";
+const DEFAULT_VALIDATORS_SUBDIR: &str = "validators";
+const DEFAULT_PRIVATE_KEY_FILENAME: &str = "private.key";
 
-impl ClientConfig {
-    /// Build a new configuration from defaults.
-    pub fn default() -> Self {
-        let data_dir = {
-            let home = dirs::home_dir().expect("Unable to determine home dir.");
-            home.join(DEFAULT_LIGHTHOUSE_DIR)
+impl ValidatorClientConfig {
+    /// Build a new configuration from defaults, which are overrided by arguments provided.
+    pub fn build_config(arguments: &ArgMatches) -> Result<Self, Error> {
+        // Use the specified datadir, or default in the home directory
+        let data_dir: PathBuf = match arguments.value_of("datadir") {
+            Some(path) => PathBuf::from(path.to_string()),
+            None => {
+                let home = dirs::home_dir().ok_or_else(|| {
+                    Error::new(ErrorKind::NotFound, "Unable to determine home directory.")
+                })?;
+                home.join(DEFAULT_VALIDATOR_DATADIR)
+            }
         };
-        fs::create_dir_all(&data_dir)
-            .unwrap_or_else(|_| panic!("Unable to create {:?}", &data_dir));
-        let server = "localhost:50051".to_string();
-        let spec = ChainSpec::foundation();
-        Self {
+        fs::create_dir_all(&data_dir)?;
+
+        let validator_dir = data_dir.join(DEFAULT_VALIDATORS_SUBDIR);
+        fs::create_dir_all(&validator_dir)?;
+
+        let server: String = match arguments.value_of("server") {
+            Some(srv) => {
+                //TODO: I don't think this parses correctly a server & port combo
+                srv.parse::<u16>()
```

Comment:
I'm not sure what the value of server is supposed to be, but I've been using the `std::net` crate. It has an `IpAddr` and a `SocketAddr`. The latter represents an ip address port combo. You can move them two and from strings. So you can do things like `srv.parse::<SocketAddr>()` to get an address and socket.

Reply:
It needs to be a `&str`, which is passed to grpc. Might just keep it like this for now, but definitely in future it would be better to parse it properly.